File: projects/animation-engine/src/animation_engine/easing.py
"""
Easing functions for smooth animation transitions.

All easing functions take a normalized time t in [0, 1] and return
an eased value, typically also in [0, 1] but some functions may
exceed this range for overshoot effects.
"""

from __future__ import annotations

import math
from typing import Callable, Dict

EasingFunction = Callable[[float], float]


# ---------------------------------------------------------------------------
# Linear
# ---------------------------------------------------------------------------

def linear(t: float) -> float:
    return t


# ---------------------------------------------------------------------------
# Quadratic
# ---------------------------------------------------------------------------

def ease_in_quad(t: float) -> float:
    return t * t


def ease_out_quad(t: float) -> float:
    return t * (2 - t)


def ease_in_out_quad(t: float) -> float:
    if t < 0.5:
        return 2 * t * t
    return -1 + (4 - 2 * t) * t


# ---------------------------------------------------------------------------
# Cubic
# ---------------------------------------------------------------------------

def ease_in_cubic(t: float) -> float:
    return t * t * t


def ease_out_cubic(t: float) -> float:
    t1 = t - 1
    return t1 * t1 * t1 + 1


def ease_in_out_cubic(t: float) -> float:
    if t < 0.5:
        return 4 * t * t * t
    t1 = 2 * t - 2
    return 0.5 * t1 * t1 * t1 + 1


# ---------------------------------------------------------------------------
# Quartic
# ---------------------------------------------------------------------------

def ease_in_quart(t: float) -> float:
    return t * t * t * t


def ease_out_quart(t: float) -> float:
    t1 = t - 1
    return 1 - t1 * t1 * t1 * t1


def ease_in_out_quart(t: float) -> float:
    if t < 0.5:
        return 8 * t * t * t * t
    t1 = t - 1
    return 1 - 8 * t1 * t1 * t1 * t1


# ---------------------------------------------------------------------------
# Quintic
# ---------------------------------------------------------------------------

def ease_in_quint(t: float) -> float:
    return t * t * t * t * t


def ease_out_quint(t: float) -> float:
    t1 = t - 1
    return 1 + t1 * t1 * t1 * t1 * t1


def ease_in_out_quint(t: float) -> float:
    if t < 0.5:
        return 16 * t * t * t * t * t
    t1 = 2 * t - 2
    return 0.5 * t1 * t1 * t1 * t1 * t1 + 1


# ---------------------------------------------------------------------------
# Sine
# ---------------------------------------------------------------------------

def ease_in_sine(t: float) -> float:
    return 1 - math.cos(t * math.pi / 2)


def ease_out_sine(t: float) -> float:
    return math.sin(t * math.pi / 2)


def ease_in_out_sine(t: float) -> float:
    return -(math.cos(math.pi * t) - 1) / 2


# ---------------------------------------------------------------------------
# Exponential
# ---------------------------------------------------------------------------

def ease_in_expo(t: float) -> float:
    if t == 0:
        return 0
    return pow(2, 10 * (t - 1))


def ease_out_expo(t: float) -> float:
    if t == 1:
        return 1
    return 1 - pow(2, -10 * t)


def ease_in_out_expo(t: float) -> float:
    if t == 0:
        return 0
    if t == 1:
        return 1
    if t < 0.5:
        return 0.5 * pow(2, 20 * t - 10)
    return 1 - 0.5 * pow(2, -20 * t + 10)


# ---------------------------------------------------------------------------
# Circular
# ---------------------------------------------------------------------------

def ease_in_circ(t: float) -> float:
    return 1 - math.sqrt(1 - t * t)


def ease_out_circ(t: float) -> float:
    return math.sqrt(1 - (t - 1) * (t - 1))


def ease_in_out_circ(t: float) -> float:
    if t < 0.5:
        return 0.5 * (1 - math.sqrt(1 - 4 * t * t))
    return 0.5 * (math.sqrt(1 - (2 * t - 2) * (2 * t - 2)) + 1)


# ---------------------------------------------------------------------------
# Elastic
# ---------------------------------------------------------------------------

def ease_in_elastic(t: float) -> float:
    if t == 0:
        return 0
    if t == 1:
        return 1
    return -pow(2, 10 * t - 10) * math.sin((t * 10 - 10.75) * (2 * math.pi) / 3)


def ease_out_elastic(t: float) -> float:
    if t == 0:
        return 0
    if t == 1:
        return 1
    return pow(2, -10 * t) * math.sin((t * 10 - 0.75) * (2 * math.pi) / 3) + 1


def ease_in_out_elastic(t: float) -> float:
    if t == 0:
        return 0
    if t == 1:
        return 1
    if t < 0.5:
        return -(pow(2, 20 * t - 10) * math.sin((20 * t - 11.125) * (2 * math.pi) / 4.5)) / 2
    return (pow(2, -20 * t + 10) * math.sin((20 * t - 11.125) * (2 * math.pi) / 4.5)) / 2 + 1


# ---------------------------------------------------------------------------
# Back
# ---------------------------------------------------------------------------

_C1 = 1.70158
_C2 = _C1 * 1.525
_C3 = _C1 + 1


def ease_in_back(t: float) -> float:
    return _C3 * t * t * t - _C1 * t * t


def ease_out_back(t: float) -> float:
    t1 = t - 1
    return 1 + _C3 * t1 * t1 * t1 + _C1 * t1 * t1


def ease_in_out_back(t: float) -> float:
    if t < 0.5:
        return (pow(2 * t, 2) * ((_C2 + 1) * 2 * t - _C2)) / 2
    return (pow(2 * t - 2, 2) * ((_C2 + 1) * (2 * t - 2) + _C2) + 2) / 2


# ---------------------------------------------------------------------------
# Bounce
# ---------------------------------------------------------------------------

def _bounce_out(t: float) -> float:
    if t < 1 / 2.75:
        return 7.5625 * t * t
    elif t < 2 / 2.75:
        t -= 1.5 / 2.75
        return 7.5625 * t * t + 0.75
    elif t < 2.5 / 2.75:
        t -= 2.25 / 2.75
        return 7.5625 * t * t + 0.9375
    else:
        t -= 2.625 / 2.75
        return 7.5625 * t * t + 0.984375


def ease_in_bounce(t: float) -> float:
    return 1 - _bounce_out(1 - t)


def ease_out_bounce(t: float) -> float:
    return _bounce_out(t)


def ease_in_out_bounce(t: float) -> float:
    if t < 0.5:
        return (1 - _bounce_out(1 - 2 * t)) / 2
    return (1 + _bounce_out(2 * t - 1)) / 2


# ---------------------------------------------------------------------------
# Registry
# ---------------------------------------------------------------------------

EASING_FUNCTIONS: Dict[str, EasingFunction] = {
    "linear": linear,
    "ease_in_quad": ease_in_quad,
    "ease_out_quad": ease_out_quad,
    "ease_in_out_quad": ease_in_out_quad,
    "ease_in_cubic": ease_in_cubic,
    "ease_out_cubic": ease_out_cubic,
    "ease_in_out_cubic": ease_in_out_cubic,
    "ease_in_quart": ease_in_quart,
    "ease_out_quart": ease_out_quart,
    "ease_in_out_quart": ease_in_out_quart,
    "ease_in_quint": ease_in_quint,
    "ease_out_quint": ease_out_quint,
    "ease_in_out_quint": ease_in_out_quint,
    "ease_in_sine": ease_in_sine,
    "ease_out_sine": ease_out_sine,
    "ease_in_out_sine": ease_in_out_sine,
    "ease_in_expo": ease_in_expo,
    "ease_out_expo": ease_out_expo,
    "ease_in_out_expo": ease_in_out_expo,
    "ease_in_circ": ease_in_circ,
    "ease_out_circ": ease_out_circ,
    "ease_in_out_circ": ease_in_out_circ,
    "ease_in_elastic": ease_in_elastic,
    "ease_out_elastic": ease_out_elastic,
    "ease_in_out_elastic": ease_in_out_elastic,
    "ease_in_back": ease_in_back,
    "ease_out_back": ease_out_back,
    "ease_in_out_back": ease_in_out_back,
    "ease_in_bounce": ease_in_bounce,
    "ease_out_bounce": ease_out_bounce,
    "ease_in_out_bounce": ease_in_out_bounce,
}


def get_easing_function(name: str) -> EasingFunction:
    """Get an easing function by name.

    Supports both underscore format (ease_in_quad) and CSS format (ease-in-quad).
    """
    # Normalize CSS-style names to underscore format
    normalized = name.replace("-", "_").lower()
    if normalized in EASING_FUNCTIONS:
        return EASING_FUNCTIONS[normalized]
    raise ValueError(f"Unknown easing function: {name!r}. "
                     f"Available: {', '.join(sorted(EASING_FUNCTIONS.keys()))}")


def list_easing_functions() -> list:
    """Return sorted list of available easing function names."""
    return sorted(EASING_FUNCTIONS.keys())

File: projects/animation-engine/src/easing.ts
/**
 * Easing Functions Library
 *
 * Provides a comprehensive set of easing functions for animations.
 * Each function takes a progress value t (0-1) and returns an eased value.
 *
 * Categories:
 * - Linear: No easing
 * - Quad/Cubic/Quart/Quint: Polynomial easing
 * - Sine: Trigonometric easing
 * - Expo: Exponential easing
 * - Circ: Circular easing
 * - Elastic: Elastic easing
 * - Back: Overshooting easing
 * - Bounce: Bouncing easing
 */

import { EasingFunction } from './types';

// ============================================================
// Linear
// ============================================================

export const linear: EasingFunction = (t) => t;

// ============================================================
// Quadratic (t^2)
// ============================================================

export const easeInQuad: EasingFunction = (t) => t * t;

export const easeOutQuad: EasingFunction = (t) => t * (2 - t);

export const easeInOutQuad: EasingFunction = (t) =>
  t < 0.5 ? 2 * t * t : -1 + (4 - 2 * t) * t;

// ============================================================
// Cubic (t^3)
// ============================================================

export const easeInCubic: EasingFunction = (t) => t * t * t;

export const easeOutCubic: EasingFunction = (t) => (--t) * t * t + 1;

export const easeInOutCubic: EasingFunction = (t) =>
  t < 0.5 ? 4 * t * t * t : (t - 1) * (2 * t - 2) * (2 * t - 2) + 1;

// ============================================================
// Quartic (t^4)
// ============================================================

export const easeInQuart: EasingFunction = (t) => t * t * t * t;

export const easeOutQuart: EasingFunction = (t) => 1 - (--t) * t * t * t;

export const easeInOutQuart: EasingFunction = (t) =>
  t < 0.5 ? 8 * t * t * t * t : 1 - 8 * (--t) * t * t * t;

// ============================================================
// Quintic (t^5)
// ============================================================

export const easeInQuint: EasingFunction = (t) => t * t * t * t * t;

export const easeOutQuint: EasingFunction = (t) => 1 + (--t) * t * t * t * t;

export const easeInOutQuint: EasingFunction = (t) =>
  t < 0.5 ? 16 * t * t * t * t * t : 1 + 16 * (--t) * t * t * t * t;

// ============================================================
// Sine
// ============================================================

export const easeInSine: EasingFunction = (t) =>
  1 - Math.cos((t * Math.PI) / 2);

export const easeOutSine: EasingFunction = (t) =>
  Math.sin((t * Math.PI) / 2);

export const easeInOutSine: EasingFunction = (t) =>
  -(Math.cos(Math.PI * t) - 1) / 2;

// ============================================================
// Exponential
// ============================================================

export const easeInExpo: EasingFunction = (t) =>
  t === 0 ? 0 : Math.pow(2, 10 * (t - 1));

export const easeOutExpo: EasingFunction = (t) =>
  t === 1 ? 1 : 1 - Math.pow(2, -10 * t);

export const easeInOutExpo: EasingFunction = (t) => {
  if (t === 0) return 0;
  if (t === 1) return 1;
  if (t < 0.5) return Math.pow(2, 20 * t - 10) / 2;
  return (2 - Math.pow(2, -20 * t + 10)) / 2;
};

// ============================================================
// Circular
// ============================================================

export const easeInCirc: EasingFunction = (t) =>
  1 - Math.sqrt(1 - t * t);

export const easeOutCirc: EasingFunction = (t) =>
  Math.sqrt(1 - (--t) * t);

export const easeInOutCirc: EasingFunction = (t) =>
  t < 0.5
    ? (1 - Math.sqrt(1 - 4 * t * t)) / 2
    : (Math.sqrt(1 - (2 * t - 2) * (2 * t - 2)) + 1) / 2;

// ============================================================
// Elastic
// ============================================================

const ELASTIC_AMPLITUDE = 1;
const ELASTIC_PERIOD = 0.3;

export const easeInElastic: EasingFunction = (t) => {
  if (t === 0 || t === 1) return t;
  const p = ELASTIC_PERIOD;
  const s = p / 4;
  return -(
    ELASTIC_AMPLITUDE *
    Math.pow(2, 10 * (t - 1)) *
    Math.sin(((t - 1 - s) * (2 * Math.PI)) / p)
  );
};

export const easeOutElastic: EasingFunction = (t) => {
  if (t === 0 || t === 1) return t;
  const p = ELASTIC_PERIOD;
  const s = p / 4;
  return (
    ELASTIC_AMPLITUDE *
    Math.pow(2, -10 * t) *
    Math.sin(((t - s) * (2 * Math.PI)) / p) +
    1
  );
};

export const easeInOutElastic: EasingFunction = (t) => {
  if (t === 0 || t === 1) return t;
  const p = ELASTIC_PERIOD * 1.5;
  const s = p / 4;
  if (t < 0.5) {
    return (
      -0.5 *
      ELASTIC_AMPLITUDE *
      Math.pow(2, 20 * t - 10) *
      Math.sin(((20 * t - 11.125) * (2 * Math.PI)) / p)
    );
  }
  return (
    ELASTIC_AMPLITUDE *
    Math.pow(2, -20 * t + 10) *
    Math.sin(((20 * t - 11.125) * (2 * Math.PI)) / p) *
    0.5 +
    1
  );
};

// ============================================================
// Back
// ============================================================

const BACK_OVERSHOOT = 1.70158;

export const easeInBack: EasingFunction = (t) =>
  t * t * ((BACK_OVERSHOOT + 1) * t - BACK_OVERSHOOT);

export const easeOutBack: EasingFunction = (t) => {
  const s = BACK_OVERSHOOT;
  return (t -= 1) * t * ((s + 1) * t + s) + 1;
};

export const easeInOutBack: EasingFunction = (t) => {
  const s = BACK_OVERSHOOT * 1.525;
  if ((t *= 2) < 1) {
    return 0.5 * (t * t * ((s + 1) * t - s));
  }
  return 0.5 * ((t -= 2) * t * ((s + 1) * t + s) + 2);
};

// ============================================================
// Bounce
// ============================================================

const bounceOut: EasingFunction = (t) => {
  const n1 = 7.5625;
  const d1 = 2.75;

  if (t < 1 / d1) {
    return n1 * t * t;
  } else if (t < 2 / d1) {
    return n1 * (t -= 1.5 / d1) * t + 0.75;
  } else if (t < 2.5 / d1) {
    return n1 * (t -= 2.25 / d1) * t + 0.9375;
  } else {
    return n1 * (t -= 2.625 / d1) * t + 0.984375;
  }
};

export const easeOutBounce: EasingFunction = bounceOut;

export const easeInBounce: EasingFunction = (t) => 1 - bounceOut(1 - t);

export const easeInOutBounce: EasingFunction = (t) =>
  t < 0.5
    ? (1 - bounceOut(1 - 2 * t)) / 2
    : (1 + bounceOut(2 * t - 1)) / 2;

// ============================================================
// Easing Registry
// ============================================================

/** Map of all named easing functions */
const easingMap: Record<string, EasingFunction> = {
  'linear': linear,
  'ease-in-quad': easeInQuad,
  'ease-out-quad': easeOutQuad,
  'ease-in-out-quad': easeInOutQuad,
  'ease-in-cubic': easeInCubic,
  'ease-out-cubic': easeOutCubic,
  'ease-in-out-cubic': easeInOutCubic,
  'ease-in-quart': easeInQuart,
  'ease-out-quart': easeOutQuart,
  'ease-in-out-quart': easeInOutQuart,
  'ease-in-quint': easeInQuint,
  'ease-out-quint': easeOutQuint,
  'ease-in-out-quint': easeInOutQuint,
  'ease-in-sine': easeInSine,
  'ease-out-sine': easeOutSine,
  'ease-in-out-sine': easeInOutSine,
  'ease-in-expo': easeInExpo,
  'ease-out-expo': easeOutExpo,
  'ease-in-out-expo': easeInOutExpo,
  'ease-in-circ': easeInCirc,
  'ease-out-circ': easeOutCirc,
  'ease-in-out-circ': easeInOutCirc,
  'ease-in-elastic': easeInElastic,
  'ease-out-elastic': easeOutElastic,
  'ease-in-out-elastic': easeInOutElastic,
  'ease-in-back': easeInBack,
  'ease-out-back': easeOutBack,
  'ease-in-out-back': easeInOutBack,
  'ease-in-bounce': easeInBounce,
  'ease-out-bounce': easeOutBounce,
  'ease-in-out-bounce': easeInOutBounce,
};

/**
 * Resolve an easing function by name or return the function directly
 */
export function resolveEasing(easing: EasingFunction | string | undefined): EasingFunction {
  if (easing === undefined) {
    return easeOutCubic;
  }
  if (typeof easing === 'function') {
    return easing;
  }
  const fn = easingMap[easing];
  if (!fn) {
    console.warn(`Unknown easing function: "${easing}", falling back to easeOutCubic`);
    return easeOutCubic;
  }
  return fn;
}

/**
 * Get list of all available easing function names
 */
export function getEasingNames(): string[] {
  return Object.keys(easingMap);
}

/**
 * Register a custom easing function
 */
export function registerEasing(name: string, fn: EasingFunction): void {
  easingMap[name] = fn;
}
